src/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -17,5 +21,136 @@ private Socket GetOrCreateAcceptSocket(Socket acceptSocket, bool unused, string
+            }
+        }
+
+        private async Task SendFileInternalAsync(FileStream fileStream, byte[] preBuffer, byte[] postBuffer)
+        {
+            SocketError errorCode = SocketError.Success;
+            using (fileStream)
+            {
+                // Send the preBuffer, if any
+                // This will throw on error
+                if (preBuffer != null && preBuffer.Length > 0)
+                {
+                    // Using "this." seems to make the extension method kick in
+                    await this.SendAsync(new ArraySegment<byte>(preBuffer), SocketFlags.None);
+                }
+
+                // Send the file, if any
+                if (fileStream != null)
+                {
+                    TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();
+
+                    // This can throw ObjectDisposedException.
+                    errorCode = SocketPal.SendFileAsync(_handle, fileStream, (bytesTransferred, socketError) => 
+                    {
+                        if (socketError != SocketError.Success)
+                        {
+                            // Synchronous exception from SendFileAsync
+                            SocketException socketException = new SocketException((int)errorCode);
+                            UpdateStatusAfterSocketError(socketException);
+                            if (NetEventSource.IsEnabled) NetEventSource.Error(this, socketException);
+                            tcs.SetException(socketException);
+                        }
+
+                        tcs.SetResult(true);
+                    });
+
+                    await tcs.Task;
+                }
+            }
+
+            if (errorCode != SocketError.Success)
+            {
+                // Synchronous exception from SendFileAsync
+                SocketException socketException = new SocketException((int)errorCode);
+                UpdateStatusAfterSocketError(socketException);
+                if (NetEventSource.IsEnabled) NetEventSource.Error(this, socketException);
+                throw socketException;
+            }
+
+            // Send the postBuffer, if any
+            // This will throw on error
+            if (postBuffer != null && postBuffer.Length > 0)
+            {
+                // Using "this." seems to make the extension method kick in
+                await this.SendAsync(new ArraySegment<byte>(postBuffer), SocketFlags.None);
+            }
+        }
+
+        private IAsyncResult BeginSendFileInternal(string fileName, byte[] preBuffer, byte[] postBuffer, TransmitFileOptions flags, AsyncCallback callback, object state)
+        {
+            // Note, UseDefaultWorkerThread is the default and is == 0.
+            // Unfortunately there is no TransmitFileOptions.None.
+            if (flags != TransmitFileOptions.UseDefaultWorkerThread)
+            {
+                throw new PlatformNotSupportedException();
+            }
```

Comment:
Same message comment as earlier.  Consider putting this into a helper that validates flags.